include/gridtools/fn/backend/gpu.hpp

```diff
@@ -23,108 +24,154 @@
-            if constexpr (ndims_t::value == 0) {
-                return hymap::keys<>::values<>();
-            } else if constexpr (ndims_t::value == 1) {
-                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
-                using values_t = typename keys_t::template values<int>;
-                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x);
-            } else if constexpr (ndims_t::value == 2) {
-                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
-                using block_dim_y = block_size_at_dim<BlockSizes, keys_t, 1>;
-                using values_t = typename keys_t::template values<int, int>;
-                return values_t(
-                    blockIdx.x * block_dim_x::value + threadIdx.x, blockIdx.y * block_dim_y::value + threadIdx.y);
-            } else {
-                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
-                using block_dim_y = block_size_at_dim<BlockSizes, keys_t, 1>;
-                using block_dim_z = block_size_at_dim<BlockSizes, keys_t, 2>;
-                using values_t = typename keys_t::template values<int, int, int>;
-                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x,
-                    blockIdx.y * block_dim_y::value + threadIdx.y,
-                    blockIdx.z * block_dim_z::value + threadIdx.z);
+        using block_sizes_for_sizes =
+            hymap::from_meta_map<meta::transform<block_size_at_dim<BlockSizes>::template apply, get_keys<Sizes>>>;
+
+        struct global_thread_index_f {
+            template <std::size_t I, class Index>
+            GT_FUNCTION_DEVICE static constexpr int index_at_dim(Index const &idx) {
+                static_assert(I < 3);
+                if constexpr (I == 0)
+                    return idx.x;
+                if constexpr (I == 1)
+                    return idx.y;
+                return idx.z;
             }
-            // disable incorrect warning "missing return statement at end of non-void function"
-            GT_NVCC_DIAG_PUSH_SUPPRESS(940)
-        }
-        GT_NVCC_DIAG_POP_SUPPRESS(940)
 
-        template <class Key>
-        struct at_generator_f {
-            template <class Value>
-            GT_FUNCTION_DEVICE decltype(auto) operator()(Value &&value) const {
-                return device::at_key<Key>(std::forward<Value>(value));
+            template <std::size_t I, class ThreadBlockSize, class LoopBlockSize>
+            GT_FUNCTION_DEVICE constexpr auto operator()(ThreadBlockSize, LoopBlockSize) const {
+                if constexpr (I < 3) {
+                    return index_at_dim<I>(blockIdx) * (ThreadBlockSize::value * LoopBlockSize::value) +
+                           index_at_dim<I>(threadIdx) * LoopBlockSize::value;
+                } else {
+                    return integral_constant<int, 0>();
```

Comment:
I am a bit lost here, maybe you can add a few comments.

Reply:
Added some at various places, let me know if that’s enough.